### DISCUSSION TITLE
Added: C# Design Notes for nullable reference types

### DISCUSSION THREAD

**Discussion by MadsTorgersen** [dotnet/csharplang#1208] [General]
I completed the outstanding design notes related to nullable reference types. Opening just one discussion issue, since some of the decisions in individual notes are outdated.

[C# Language Design Notes for Aug 16, 2017](https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-08-16.md)

1. The null-forgiving operator

[C# Language Design Notes for Aug 23, 2017](https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-08-23.md)

We discussed various aspects of nullable reference types

1. How does flow analysis silence the warning
2. Problems with dotted names
3. Type inference
4. Structs with fields of non-nullable type

[C# Language Design Review, Oct 4, 2017](https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-10-04.md)
_Added earlier. Included here for completeness._

We looked at nullable reference types with the reviewers, Anders Hejlsberg and Kevin Pilch.

1. Overall philosophy
2. Switches
3. Libraries
4. Dotted names
5. Type narrowing
6. The dammit operator
7. Array covariance
8. Null warnings
9. Special methods
10. Conclusion

[C# Language Design Notes for Oct 11, 2017](https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-10-11.md)

We looked at the Oct 4 design review feedback for nullable reference types, and considered how to react to it.

1. Philosophy
2. Switches
3. Dotted names
4. Type narrowing
5. Dammit operator type narrowing
6. Dammit operator stickiness
7. Array covariance
8. Null warnings

[C# Language Design Notes for Nov 8, 2017](https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-11-08.md)

We went over the status of the prototype for nullable reference types, to address outstanding questions and make any last minute calls before release.

1. Constructors
2. Dotted names
3. Default expressions
4. Should we track null state for nonnullable ref types?
5. Inferred types for method type inference
6. Inferred nullability in hover tips
7. Smaller things not yet done
8. Unconstrained generics
9. Other issues

[C# Language Design Notes for Nov 27, 2017](https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-11-27.md)

We went over the feedback on the nullable reference types prototype, and discussed how to address the top issues that people had found using the feature on their own source code.

1. Interacting with existing, unannotated APIs
2. Accommodating alternative initialization patterns
3. Tracking nullable value types
4. Tracking dotted names
5. Special methods
6. Filtering out nulls

**Comment by orthoxerox** [comment]
> Interacting with existing, unannotated APIs

Have you thought about adding support for something similar to `.d.ts` in TypeScript? It's improbable that every single version of every single library in existence used by someone else will be recompiled using C#8 and reuploaded to NuGet or other distribution channels.

But files similar to Roslyn's PublicAPI.txt could be sideloaded into existing projects to mark the nullability of existing libraries. They could even be crowdsourced for libraries that are no longer supported by the creators or where the source code of a specific version is lost.

**Comment by iam3yal** [comment]
@orthoxerox `.d.cs` can also be a good approach to introduce DbC into the language in the future or even things like documentation.

**Comment by yaakov-h** [comment]
FWIW, Code Contracts uses reference assemblies for this purpose - given `path\to\Foo.dll`, Code Contracts looks for `path\to\CodeContracts\Foo.dll` and reads the contracts from there. This allows anyone to write contracts for pre-existing assemblies.

Additionally, it also provides Contract reference assemblies for the BCL.

Even if C#8 goes a different route and doesn't use reference assemblies to annotate nullability, it might be worth piggybacking off of Code Contracts and converting the nullability portions of the BCL contracts into whatever form C#8 uses to annotate pre-existing assemblies.

**Comment by iam3yal** [comment]
> Dammit operator stickiness

> M(s!, s);                           // why not here?

Because this might be too confusing, it's not scoped and people will have to play the analyzer and figure whether it's nullable or non-nullable based on the flow of the code so being explicit about makes the code easier to follow.

To make it stick all people have to do is wrap the section of code with an `if` statement or whatever and check against null so the trade off isn't too bad.

```
string? s = null;

if (s != null) {
	if (s.Length == 5) {
		// ...
	}

	M(s, s); // s is string.
}

M(s!, s); // s is string?.
```

**Comment by svick** [comment]
> Filtering out nulls

Would it make sense to consider `Where` to be one of the special methods, that ensures the given `null`-related conditions hold for each item in the collection?

For example, consider:

```c#
class Phone {}

class Person
{
    string Name { get; }
    Phone? Phone { get; }
}

void Call(string name, Phone phone) { ... }

void CallEveryone(IEnumerable<Person> people)
{
    var peopleWithPhones = people.Where(p => p.Phone != null);

    foreach (var person in peopleWithPhone)
        Call(person.Name, person.Phone);
}
```

`WhereNotNull()` wouldn't help with this code. Or is there some other way to make this code `null`-safe
according to the compiler while using LINQ and without using `!`?

**Comment by svick** [comment]
>  Arrays are (unsafely) covariant over reference types, and for consistency we should also make them covariant over nullability.

Why? Array covariance is a bad feature and consistency with a bad feature does not seem like something to strive for to me.

If I really want to make the conversion, I can do it by using `!`. But I don't see why it should be allowed otherwise.

**Comment by dimka11** [comment]
How can I install Nullable preview, if I don't have preview version of Visual Studio and "Installing on a machine with multiple versions of Visual Studio is not supported."? I have to completly unistall VS 2017, then install VS 2017 Preview? It's will be too long, with all C++, Python, SDK's and other stuff.

**Comment by yaakov-h** [comment]
This is off-topic for design discussions, however, you only need 15.5, which is no longer in preview.

I wouldn't recommend installing an experimental compiler in your production instance of Visual Studio. Personally, I'd recommend using a virtual machine.

**Comment by lucasteles** [comment]
Have we some idea if we will be able to create extensions for nullable reference types? 

today `Nullable<T>` has a constraint forcing T to be a struct, 
so we cant create some 'monad flow'  for a  type `T?`, what woud be great

**Comment by qrli** [comment]
> Interacting with existing, unannotated APIs

I would like to see nullability annotations to be supported and released first, so that annotated APIs will be widespread, before flow analysis reach mature state.

The most useful part is the function signiture annotation, which can be used by documentation and intellisense, while even local variable annotation could be out of MVP. 

Tools like Resharper can already do some simple nullability analysis today. The analysis does not have to be perfect and stable as carved in stone; it has a lot of room to improve in the long run. A progressive approach across several releases fits better IMO.

**Comment by Joe4evr** [comment]
@lucasteles 
> Have we some idea if we will be able to create extensions for nullable reference types?

There is no difference between a reference type and its "nullable" version at runtime, this feature is only compile-time annotations. Read my comment [over here](https://github.com/dotnet/csharplang/issues/36#issuecomment-307543108) which notes the relevant parts.
